client/client.go

```diff
@@ -22,57 +22,334 @@ limitations under the License.
 package client
 
 import (
+	"errors"
 	"fmt"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"sync"
 
 	"github.com/fsouza/go-dockerclient"
+	"github.com/intelsdi-x/snap-plugin-collector-docker/config"
+	"github.com/intelsdi-x/snap-plugin-collector-docker/fs"
+	"github.com/intelsdi-x/snap-plugin-collector-docker/network"
+	"github.com/intelsdi-x/snap-plugin-collector-docker/wrapper"
 	"github.com/opencontainers/runc/libcontainer/cgroups"
 )
 
-const endpoint string = "unix:///var/run/docker.sock"
+const (
+	endpoint         string = "unix:///var/run/docker.sock"
+	dockerVersionKey string = "Version"
+)
 
-type ContainerInfo struct {
-	Id      string
-	Status  string
-	Created int64
-	Image   string
+// DockerClientInterface provides methods i.a. for interaction with the docker API.
+type DockerClientInterface interface {
+	ListContainersAsMap() (map[string]docker.APIContainers, error)
+	GetStatsFromContainer(string, bool) (*wrapper.Statistics, error)
+	InspectContainer(string) (*docker.Container, error)
+	FindCgroupMountpoint(string) (string, error)
 }
 
-type DockerClientInterface interface {
-	ListContainers() ([]ContainerInfo, error)
-	FindCgroupMountpoint(subsystem string) (string, error)
+// DockerClient
+type DockerClient struct {
+	cl                 *docker.Client
+	inspectCache       map[string]*docker.Container
+	inspectMutex       sync.Mutex
+	diskUsageCollector fs.DiskUsageCollector
 }
 
-type dockerClient struct{}
+type deviceInfo struct {
+	device string
+	major  string
+	minor  string
+}
 
-func NewDockerClient() *dockerClient {
-	return new(dockerClient)
+// NewDockerClient returns dockerClient instance ready for communication with the server endpoint `unix:///var/run/docker.sock`
+func NewDockerClient() (*DockerClient, error) {
+	client, err := docker.NewClient(endpoint)
+	if err != nil {
+		return nil, fmt.Errorf("Cannot initialize docker client instance with the given server endpoint `%s`, err=%v", endpoint, err)
+	}
+
+	dc := &DockerClient{
+		cl:                 client,
+		inspectCache:       map[string]*docker.Container{},
+		diskUsageCollector: fs.DiskUsageCollector{},
+	}
+
+	dc.diskUsageCollector.Init()
+	// get version of docker engine
+	version, err := dc.version()
+	if err != nil {
+		return nil, err
```

Comment:
What do you think about logging at least error cases, maybe that would simplify debugging in deployment environments?